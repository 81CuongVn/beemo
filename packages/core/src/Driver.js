/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Plugin } from 'boost';
import merge from 'lodash/merge';
import Options, { array, bool, number, object, shape, string, union } from 'optimal';

import type { EventListener } from 'boost';
import type {
  DriverCommandOptions,
  DriverContext,
  DriverOptions,
  DriverMetadata,
  Execution,
} from './types';

export default class Driver extends Plugin<DriverOptions> {
  command: DriverCommandOptions = {};

  context: DriverContext;

  metadata: DriverMetadata;

  constructor(options?: $Shape<DriverOptions> = {}) {
    super(options);

    this.options = new Options(
      options,
      {
        args: array(string()),
        dependencies: array(string()),
        env: object(union([bool(), number(), string()])),
      },
      {
        name: this.constructor.name,
      },
    );
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: Object): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): string[] {
    return [...this.options.args];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...this.options.dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Handle command failures according to this driver.
   */
  handleFailure(error: Execution) {
    const { stderr, stdout } = error;
    const out = (stderr || stdout).trim();

    // Integration debugging
    // this.tool.logError(JSON.stringify(error));

    if (out) {
      this.tool.logError(out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  handleSuccess(response: Execution) {
    const out = response.stdout.trim();

    // Integration debugging
    // this.tool.log(JSON.stringify(response));

    if (out) {
      this.tool.log(out);
    }
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Object, next: Object): Object {
    return merge(prev, next);
  }

  /**
   * Easily register events in the tool.
   */
  on(eventName: string, listener: EventListener): this {
    this.tool.on(eventName, listener);

    return this;
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint = {};

    Object.keys(options).forEach(key => {
      blueprint[key] = shape({
        alias: union([string(), array(string())], ''),
        description: string().required(),
      });
    });

    this.command = new Options(options, blueprint, {
      name: this.constructor.name,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: $Shape<DriverMetadata>): this {
    this.metadata = new Options(
      metadata,
      {
        bin: string()
          .match(/^[-a-z0-9]+$/)
          .required(),
        configName: string().required(),
        configOption: string('--config').empty(),
        dependencies: array(string()),
        description: string().empty(),
        filterOptions: bool(false),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }
}
