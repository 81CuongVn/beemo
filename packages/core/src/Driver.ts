/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Plugin, EventListener } from 'boost';
import merge from 'lodash/merge';
import optimal, { array, bool, number, object, shape, string, union, Blueprint } from 'optimal';
import DriverContext from './contexts/DriverContext';
import {
  Argv,
  BeemoConfig,
  DriverCommandOptions,
  DriverOptions,
  DriverMetadata,
  Execution,
} from './types';

export const STRATEGY_REFERENCE = 'reference';
export const STRATEGY_COPY = 'copy';

export default class Driver<T> extends Plugin<DriverOptions> {
  command: DriverCommandOptions = {};

  // @ts-ignore Set after instantiation
  config: T;

  // @ts-ignore Set after instantiation
  context: DriverContext;

  // @ts-ignore Set after instantiation
  metadata: DriverMetadata;

  constructor(options: Partial<DriverOptions> = {}) {
    super(options);

    this.options = optimal(
      options,
      {
        args: array(string()),
        dependencies: array(string()),
        env: object(union([bool(), number(), string()])),
      },
      {
        name: this.constructor.name,
      },
    );
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: T): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return [...this.options.args];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...this.options.dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Handle command failures according to this driver.
   */
  handleFailure(error: Execution) {
    const { stderr, stdout } = error;
    const out = (stderr || stdout).trim();

    // Integration debugging
    // this.tool.logError(JSON.stringify(error));

    if (out) {
      this.tool.logError(out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  handleSuccess(response: Execution) {
    const out = response.stdout.trim();

    // Integration debugging
    // this.tool.log(JSON.stringify(response));

    if (out) {
      this.tool.log(out);
    }
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: T, next: T): T {
    return merge(prev, next);
  }

  /**
   * Easily register events in the tool.
   */
  on(eventName: string, listener: EventListener): this {
    this.tool.on(eventName, listener);

    return this;
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint = {};

    Object.keys(options).forEach(key => {
      blueprint[key] = shape({
        alias: union([string(), array(string())], ''),
        description: string().required(),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[-a-z0-9]+$/)
          .required(),
        configName: string().required(),
        configOption: string('--config').empty(),
        dependencies: array(string()),
        description: string().empty(),
        filterOptions: bool(false),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }
}
