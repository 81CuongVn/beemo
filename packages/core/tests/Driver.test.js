import { Tool } from 'boost';
import Driver from '../src/Driver';
import { createDriver } from '../../../tests/helpers';

jest.mock('boost/lib/Tool');

describe('Driver', () => {
  let driver;

  beforeEach(() => {
    driver = createDriver('foo', new Tool());
  });

  it('validates options', () => {
    expect(() => {
      driver = new Driver({
        args: true,
      });
    }).toThrowError('Invalid Driver option "args". Must be an array.');
  });

  describe('formatConfig()', () => {
    it('formats to JSON', () => {
      expect(
        driver.formatConfig({
          foo: 123,
          bar: 'abc',
          baz: true,
        }),
      ).toMatchSnapshot();
    });
  });

  describe('getArgs()', () => {
    it('returns passed options', () => {
      driver.options.args = ['--foo'];

      expect(driver.getArgs()).toEqual(['--foo']);
    });
  });

  describe('getDependencies()', () => {
    it('returns both option and metadata dependencies', () => {
      driver.metadata.dependencies = ['foo'];
      driver.options.dependencies = ['bar'];

      expect(driver.getDependencies()).toEqual(['foo', 'bar']);
    });
  });

  describe('getSupportedOptions()', () => {
    it('returns an array of options', () => {
      expect(driver.getSupportedOptions()).toEqual([]);
    });
  });

  describe('handleFailure()', () => {
    it('logs stdout', () => {
      driver.handleFailure({
        stdout: 'out',
      });

      expect(driver.tool.logError).toHaveBeenCalledWith('out');
    });

    it('logs stderr', () => {
      driver.handleFailure({
        stderr: 'error',
      });

      expect(driver.tool.logError).toHaveBeenCalledWith('error');
    });

    it('logs stderr over stdout', () => {
      driver.handleFailure({
        stderr: 'error',
        stdout: 'out',
      });

      expect(driver.tool.logError).toHaveBeenCalledWith('error');
    });

    it('doesnt log if empty', () => {
      driver.handleFailure({
        stderr: '',
        stdout: '',
      });

      expect(driver.tool.logError).not.toHaveBeenCalled();
    });
  });

  describe('handleSuccess()', () => {
    it('logs stdout', () => {
      driver.handleSuccess({
        stdout: 'out',
      });

      expect(driver.tool.log).toHaveBeenCalledWith('out');
    });

    it('doesnt log stdout if empty', () => {
      driver.handleSuccess({
        stdout: '',
      });

      expect(driver.tool.log).not.toHaveBeenCalledWith();
    });

    it('doesnt log stderr', () => {
      driver.handleFailure({
        stderr: 'error',
      });

      expect(driver.tool.log).not.toHaveBeenCalled();
    });

    it('doesnt log if empty', () => {
      driver.handleFailure({
        stderr: '',
        stdout: '',
      });

      expect(driver.tool.log).not.toHaveBeenCalled();
    });
  });

  describe('mergeConfig()', () => {
    it('deep merges objects', () => {
      expect(
        driver.mergeConfig(
          {
            foo: 123,
            bar: {
              baz: 'abc',
            },
          },
          {
            bar: {
              baz: 'xyz',
              qux: true,
            },
          },
        ),
      ).toEqual({
        foo: 123,
        bar: {
          baz: 'xyz',
          qux: true,
        },
      });
    });
  });

  describe('on()', () => {
    it('pipes event to tool', () => {
      const func = () => {};

      driver.on('foo', func);

      expect(driver.tool.on).toHaveBeenCalledWith('foo', func);
    });
  });

  describe('setCommandOptions()', () => {
    it('supports alias as a string', () => {
      driver.setCommandOptions({
        foo: {
          alias: 'f',
          description: 'Foo',
        },
      });

      expect(driver.command).toEqual({
        foo: {
          alias: 'f',
          description: 'Foo',
        },
      });
    });

    it('supports alias as an array of strings', () => {
      driver.setCommandOptions({
        foo: {
          alias: ['f', 'o'],
          description: 'Foo',
        },
      });

      expect(driver.command).toEqual({
        foo: {
          alias: ['f', 'o'],
          description: 'Foo',
        },
      });
    });

    it('doesnt support alias as other types', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {
            alias: 123,
          },
        });
      }).toThrowError(
        'Invalid Driver option "foo.alias". Type must be one of: String, Array<String>',
      );
    });

    it('supports description as a string', () => {
      driver.setCommandOptions({
        foo: {
          description: 'Foo',
        },
      });

      expect(driver.command).toEqual({
        foo: {
          description: 'Foo',
        },
      });
    });

    it('doesnt support description as other types', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {
            description: 123,
          },
        });
      }).toThrowError('Invalid Driver option "foo.description". Must be a string.');
    });

    it('doesnt support empty descriptions', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {
            description: '',
          },
        });
      }).toThrowError('Invalid Driver option "foo.description". String cannot be empty.');
    });

    it('requires a description', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {},
        });
      }).toThrowError(
        'Invalid Driver option "foo.description". Field is required and must be defined.',
      );
    });
  });

  describe('setMetadata()', () => {
    const options = {
      bin: 'beemo',
      configName: 'beemo',
      title: 'Beemo',
    };

    describe('bin', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({});
        }).toThrowError('Invalid Driver option "bin". Field is required and must be defined.');
      });

      it('supports alpha, numeric, and dashes', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'foo-123',
          });
        }).not.toThrowError();
      });

      it('doesnt support underscores', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'foo_123',
          });
        }).toThrowError(
          'Invalid Driver option "bin". String does not match pattern "^[-a-z0-9]+$".',
        );
      });

      it('doesnt support uppercase', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'Foo_123',
          });
        }).toThrowError(
          'Invalid Driver option "bin". String does not match pattern "^[-a-z0-9]+$".',
        );
      });
    });

    describe('configName', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({
            bin: 'beemo',
          });
        }).toThrowError(
          'Invalid Driver option "configName". Field is required and must be defined.',
        );
      });

      it('doesnt support non-strings', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            configName: 123,
          });
        }).toThrowError('Invalid Driver option "configName". Must be a string.');
      });
    });

    describe('title', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({
            bin: 'beemo',
            configName: 'beemo',
          });
        }).toThrowError('Invalid Driver option "title". Field is required and must be defined.');
      });

      it('doesnt support non-strings', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            title: 123,
          });
        }).toThrowError('Invalid Driver option "title". Must be a string.');
      });
    });
  });
});
