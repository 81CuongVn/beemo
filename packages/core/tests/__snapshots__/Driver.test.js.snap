// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Driver formatConfig() formats to JS if config ends in .js 1`] = `
"module.exports = {
  \\"foo\\": 123,
  \\"bar\\": \\"abc\\",
  \\"baz\\": true
};"
`;

exports[`Driver formatConfig() formats to JSON if config ends in .json 1`] = `
"{
  \\"foo\\": 123,
  \\"bar\\": \\"abc\\",
  \\"baz\\": true
}"
`;
